source-code/core/src/lint/RFC.md

```diff
@@ -0,0 +1,128 @@
+
+3. What information would a lint error emit?
+
+	```ts
+	type LintInformation = {
+		type: LintRule // which rule emitted the error
+		message: string // a string that informs what is wrong
+	}
+	```
+	Question: Anything else?
+
+	Question: Should any lint rule be able to extend this object?
+
+<!------------------------------------------------------------------------------------------------>
+
+## Configure Lint behavior
+
+`inlang` will provide some lint functionality that can be configured or disabled.
+It does not need to be as complicated as eslint, but it should follow the same design when it comes to the `rules`.
+The `inlang.config.js` will be extended with a new `linting` property.
+
+```js
+/**
+ * @type {import("@inlang/core/config").DefineConfig}
+ */
+export async function defineConfig(env) {
+  	return {
+   	referenceLanguage: 'en,
+   	languages: ['en', 'de'],
+		linting: {
+			missing_key: false, // disable this specific lint rule
+			missing_key: 'warn', // override the level
+			missing_key: { 'error', { ignore: ['test-key'] }}, // configure the linting behavior where needed
+		}
+  	}
+}
+```
+
+See [`experiment-3.ts`](https://github.com/ivanhofer/inlang/blob/lint/source-code/core/src/lint/experiment-3.ts) for more details.
+
+Because we are in control of all lint rules, we can have strong type support for all the lint configurations.
+
+### extending lint rules
+
+How do we want to deal with this?
+
+Question: Should anyone be able to write and add custom lint rules? _Probably yes._
+The alternative would be that someone has to create a PR and submit a new rule to the `inlang` repo. This would not be ideal as it introduces a hurdle to test something out quickly or someone might not be allowed to share his code or it may take a few days/weeks until the PR get's merged.
+
+If we allow anyone to add custom rules, we need to extend the `linting` prop.
+
+```js
+{
+	linting: {
+		plugins: [myCustomLintPlugin],
+		rules: {
+			missing_key: false, // disable this specific lint rule
+		}
+	}
+}
+```
+
+Where `myCustomPlugin` is a `JavaScript` function which needs to implement a certain interface (not yet defined).
+
+Question: If no plugin get's specified do we want to allow to get rid of the `rules` level in the config?
+
+```js
+{
+	linting: {
+		rules: {
+			missing_key: false, // disable this specific lint rule
+		}
+	}
+}
+```
+
+vs.
+
+
+```js
+{
+	linting: {
+		missing_key: false, // disable this specific lint rule
+	}
+}
+```
+
+<!------------------------------------------------------------------------------------------------>
+
+## dealing with lint errors/warnings
+
+The `@inlang/core/lint` package should provide useful utility functions that make it possible to traverse a `Resource` and check its linting properties.
```

Comment:
`resource.lint` will just contain the errors and warnings of the Resource node itself and not it's children. If we would check for `hasErrors` then we would not check for the errors of the children.

Reply:
And it would be a small memory overhead if we store additional attributes that can be easily calculated on the fly when needed.

Reply:
I see. For me, a resource's lints are referring to the children (messages). Thus, resource.hasErrors = messages in the resource have errors. 

I mean utility functions are fine and can be extended with a caching layer in the future. So, let's go with utility functions. It would be cool if @inlang/core/query could be extended with those utility function in the future. But let's keep it simple for now.

Reply:
It would be confusing to have `hasErrors` set to true and then accessing the `errors` property and get back an empty array, because just one of the children contains an error. The utility functions are the better approach.